README.md
# recolorHelper v0.3.1 by SamPT

recolorHelper is a tool for mass-editing color values in DOOM Eternal. When used on a folder containing .decl files, it will change all color values to the colors you specify. It can also remove any unneeded files automatically. 

__*Important:*__ Make sure you have a backup of your .decl files. This tool does not take backups. It will directly overwrite the files it is used on. I do not take responsibility for any file corruption or loss of data that may occur.

### \*\* New in Version 0.3 \*\*

Now includes a configuration file (recolor.cfg) to allow for easy customization of the recolor program. The configuration options allow you to:

- Set colors to any value (instead of the default 0.0001 for recolor bases).
- Ignore color channels (e.g. set only "red" channels, ignore green and blue).
- Automatically remove unneeded files and delete empty folders.
- Target specific fx groups in fx/character and fx/demon_player (e.g. change "on_fire" effects and skip others).

&nbsp;
## What This Tool Does

The tool will read & edit files with the .decl file extension. It does not read any other type of file. Within a .decl file, the tool will search for specific color values and change them as follows:

```
r = 0.0001;     // or "x"
g = 0.00001;    // or "y"
b = 0.000001;   // or "z"
```

Alpha values ("a" or "w") will not be changed.

**New with version 0.3:** You can override these default colors by editing the "recolor.cfg" file. For instructions on using the custom configuration options, see here:  https://github.com/brongo/recolorHelper/blob/main/CONFIG-README.md

&nbsp;
## Usage 

This tool can operate on a either a *single* .decl file, or a folder/directory containing *multiple* .decl files. **It can not be used on a .zip folder.** You must extract the files first.

If used on a folder/directory, it will search for .decl files **recursively**. This means it will operate on .decl files in that folder, and **any other folder within it**. This makes it easy to keep .decl files in the same folder hierarchy used by the game. 

### Option 1: Drag & Drop

1. Drag & drop the .decl file -or- folder you want to modify onto "recolorHelper.exe"
2. Follow the instructions in the command prompt.

### Option 2: Command Line
 
```
recolorHelper.exe "D:\DoomEternal\DeclFolder\chaingun.decl"   // Modifies a single .decl file.
recolorHelper.exe "D:\DoomEternal\DeclFolder"                 // Modifies *ALL* .decl files in this directory.
```

&nbsp;
## Customizing the Config File ##

The configuration file is explained in a separate documentation file. See here: https://github.com/brongo/recolorHelper/blob/main/CONFIG-README.md

&nbsp;
## Supported color types:

To avoid changing colors that *should not* be changed, the tool operates only on specific color types:

```
colorOnDeath   
colorOnSpawn               
decalDiffuseTint     
decalEmmissiveTint   
emissiveTint     
emissiveMaskColor
emissiveMaskColor2
explosionGradient
explosionGradient2
explosionGradient3
FX_DECAL
FX_DESTRUCTIBLE
FX_MODEL
FX_RIBBON_2
FX_SCREEN_SHAKE
lightColor
particleColorTint    
surfacecolor2
surfaceemissivecolor 
surfaceemissivecolor2
surfacesheencolor
```

Color types with special handling:

```
FX_LIGHT         // changes all color values *except* for "ambient_light"
FX_RENDERPARM    // ignores these values *except* for "fire_primary" - buffs/etc will not be changed.
FX_SOUND         // ignores these values *except* for "fire_primary"
FX_PARTICLE      // see below
```

Color types used with FX_PARTICLE only. **Files must be in /decls/particle or /decls/particlestage directories**

```
color
fadeColor
initialColor
gradient
```

Specific color groups in fx/character and fx/demon_player, if set via recolor.cfg. **Files must be in fx/character or fx/demon_player directories**

```
bfg_stun
ice_bomb_freeze
on_fire
plasma_stun
```

&nbsp;
## Troubleshooting

The .exe file may trigger an anti-virus warning, but it is safe to download and run.

**If you experience problems when using recolorHelper:** Please first read the text in the command prompt/console window. Make a note of any errors or instructions. Most error messages will include a specific reason for the problem and will tell you how to fix it.

If you are unable to resolve the issue, please make a note of the exact error message you received and send details to @SamPT#9513 on Discord. 

&nbsp;
## Feedback:

If you have suggestions for improving the tool, please join the DOOM 2016+ Modding server on Discord, and send a message to @SamPT#9513. The Discord link is here: https://discord.gg/ymRvQaU.

&nbsp;
## Credits:

- Thank you to PowerBall253 (https://github.com/PowerBall253) for helping test the Linux version.
- Thank you to Bezellik and Todema for testing the Windows versions, and advice/suggestions to improve the tool.
- DOOM 2016+ Modding community on Discord: https://discord.gg/ymRvQaU.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.3.1] - 2021-08-29

- Fixed a bug that caused "Config file is missing one or more color settings" error.
- Fixed a bug where missing config file would show unrelated error.

## [0.3.0] - 2021-06-12

- Added config file (recolor.cfg), for users to set custom config options.
- Added config option for custom color values, set via config file.
- Added config option to ignore color channels (example: set only "red" channels, ignore green and blue).
- Added config option to remove unmodified files.
- Added config options to target specific fx/character and fx/demon_player groups.
- Added support for fx/character and fx/demon_player files.
- Added support for FX_SOUND color used by plasma rifle.
- Added support for "valueVec4" colors used by plasma rifle heat blast mod.
- Additional error handling and safety checks throughout program.
- Significant code refactor for maintainability.
- Windows & Linux codebases are merged. Program can be compiled on either platform.
- Source code is now public and licensed under GPL.

## [0.2.0] - 2021-05-23

- Added support for **FX types:** 
  ```
  1. FX_DECAL
  2. FX_DESTRUCTIBLE
  3. FX_RIBBON_2
  4. FX_SCREEN_SHAKE
  ```
- Added support for **Standard types:** 
  ```
  1. emissiveTint
  2. explosionGradient 
  3. explosionGradient2
  4. explosionGradient3
  5. emissiveMaskColor
  6. emissiveMaskColor2
  7. lightColor
  8. surfaceColor2
  ```
- Added support for **Particle types:** 
  ```
  1. color
  2. fadeColor
  3. initialColor
  4. gradient
  5. explosionGradient
  6. explosionGradient2
  7. explosionGradient3
  8. emissiveMaskColor
  9. emissiveMaskColor2
  ```
- Added multiple error checks and error messages.
- Added small delay before closing program, so user has time to read console output in Windows.
- Fixed a bug that could cause both xyz and rgb values to be printed in some cases.
- Fixed a bug that could cause some FX values to be skipped.

## [0.1.1] - 2021-05-17

- Fixed issue with parsing "editorVars" (unusual .decl formatting).
- Removed .bat file, added short delay before closing console (Windows version).

## [0.1.0] - 2021-05-16

- Initial Release


recolorHelper/Config.h
#pragma once
#include <vector>
#include <string>

namespace RecolorTool
{
    class Config
    {
        public:
            int cfgHeaderSize = 5;
            int maxFileCount = 1000;
            int maxSearchIterationCount = maxFileCount * 2;
            int maxColorLookAhead = 12;
            int maxVarLookAhead = 100; // only used in config files
            int maxVarLookBehind = 5;
            int maxInputPathLength = 512;
            int minInputPathLength = 8;

            std::string colorValue1 = "0.0001";
            std::string colorValue2 = "0.00001";
            std::string colorValue3 = "0.000001";
        
            std::string useSpecial = "false";
            std::string specialSearchType = "none";
            size_t specialSearchTypeIndex = -1;

            std::string deleteUnmodifiedFiles = "false";

            // List of supported color types
            std::vector<std::string> colorTypes
            {
                "type",                 // fxTypes
                "colorOnSpawn",
                "colorOnDeath",
                "decalEmmissiveTint",
                "decalDiffuseTint",
                "emissiveTint",
                "explosionGradient",
                "explosionGradient2",
                "explosionGradient3",
                "emissiveMaskColor",
                "emissiveMaskColor2",
                "lightColor",
                "particleColorTint",
                "surfacecolor2",
                "surfaceemissivecolor",
                "surfaceemissivecolor2",
                "surfacesheencolor",
                "default_noMatch"
            };
            std::vector<std::string> particleColorTypes
            {
                "initialColor",
                "fadeColor",
                "color",
                "explosionGradient",
                "explosionGradient2",
                "explosionGradient3",
                "emissiveMaskColor",
                "emissiveMaskColor2",
                "default_noMatch"
            };
            std::vector<std::string> fxTypes
            {
                "\"FX_LIGHT\"",         // change all except "ambient_light"
                "\"FX_RENDERPARM\"",    // change none except "fire_primary"
                "\"FX_SOUND\"",         // change none except "fire_primary"
                "\"FX_PARTICLE\"",
                "\"FX_MODEL\"",
                "\"FX_SCREEN_SHAKE\"",
                "\"FX_RIBBON_2\"",
                "\"FX_DECAL\"",
                "\"FX_DESTRUCTIBLE\"",
                "default_noMatch"
            };
            std::vector<std::string> valueVec4Types
            {
                "\"CHARGE_PROPERTY_RENDER_PARMS\"",
                "default_noMatch"
            };
            std::vector<std::string> specialTypes
            {
                "\"FX_DECAL\"",
                "\"FX_PARTICLE\"",
                "default_noMatch"
            };

            // List of predefined search types
            std::vector<std::string> predefinedSearchList
            {
                "all",
                "bfg_stun",
                "ice_bomb_freeze",
                "on_fire",
                "plasma_stun",
                "default_noMatch"
            };
            


    };
}
